lib/semantics/resolve-names.cc

```diff
@@ -1037,11 +1037,271 @@ class ConstructVisitor : public DeclarationVisitor {
   void PopAssociation();
 };
 
+static const parser::Name *GetDesignatorNameIf(
+    const parser::Designator &designator) {
+  const auto *dataRef{std::get_if<parser::DataRef>(&designator.u)};
+  return dataRef ? std::get_if<parser::Name>(&dataRef->u) : nullptr;
+}
+
+static constexpr Symbol::Flags dataSharingAttributeFlags{
+    Symbol::Flag::OmpShared, Symbol::Flag::OmpPrivate,
+    Symbol::Flag::OmpFirstPrivate, Symbol::Flag::OmpLastPrivate,
+    Symbol::Flag::OmpReduction, Symbol::Flag::OmpLinear};
+
+static constexpr Symbol::Flags ompFlagsRequireNewSymbol{
+    Symbol::Flag::OmpPrivate, Symbol::Flag::OmpLinear,
+    Symbol::Flag::OmpFirstPrivate, Symbol::Flag::OmpLastPrivate,
+    Symbol::Flag::OmpReduction};
+
+static constexpr Symbol::Flags ompFlagsRequireMark{
+    Symbol::Flag::OmpThreadprivate};
+
+// Resolve OpenMP construct entities and statement (TODO) entities
+class OmpVisitor : public virtual DeclarationVisitor {
+public:
+  bool Pre(const parser::OpenMPBlockConstruct &) {
+    PushScope(Scope::Kind::Block, nullptr);
+    return true;
+  }
+  void Post(const parser::OpenMPBlockConstruct &) { PopScope(); }
+  bool Pre(const parser::OmpBeginBlockDirective &) {
+    clear_dataSharingAttributeObject();
+    return true;
+  }
+  void Post(const parser::OmpBeginBlockDirective &) {
+    clear_dataSharingAttributeObject();
```

Comment:
Why do you have to clear in both `Pre` and `Post`. I would have thought one of them would be enough.